clients/lsp-csharp.el

```diff
@@ -235,6 +241,188 @@ tarball or a zip file (based on a current platform) to TARGET-DIR."
   "Resolves path and arguments to use to start the server."
   (list (lsp-csharp--language-server-path) "-lsp"))
 
+(defun lsp-csharp--get-buffer-code-elements ()
+  "Retrieve code structure by calling into the /v2/codestructure endpoint.
+Returns :elements from omnisharp:CodeStructureResponse."
+  (-let* ((code-structure (lsp-request "o#/v2/codestructure"
+                                       (lsp-make-omnisharp-code-structure-request :file-name (buffer-file-name))))
+          ((&omnisharp:CodeStructureResponse :elements) code-structure))
+    elements))
+
+(defun lsp-csharp--inspect-code-elements-recursively (fn elements)
+  "Invoke FN for every omnisharp:CodeElement found recursively in ELEMENTS."
+  (seq-each
+   (lambda (el)
+     (funcall fn el)
+     (-let (((&omnisharp:CodeElement :children) el))
+       (lsp-csharp--inspect-code-elements-recursively fn children)))
+   elements))
+
+(defun lsp-csharp--collect-code-elements-recursively (predicate elements)
+  "Flatten the omnisharp:CodeElement tree in ELEMENTS matching PREDICATE."
+  (let ((results nil))
+    (lsp-csharp--inspect-code-elements-recursively (lambda (el)
+                                                     (if (funcall predicate el)
+                                                         (setq results (cons el results))))
+                                                   elements)
+    results))
+
+(lsp-defun lsp-csharp--l-c-within-range (l c (&omnisharp:Range :start :end))
+  "Determine if L (line) and C (column) are within RANGE."
+  (-let* (((&omnisharp:Point :line start-l :column start-c) start)
+          ((&omnisharp:Point :line end-l :column end-c) end))
+    (or (and (= l start-l) (>= c start-c) (or (> end-l start-l) (<= c end-c)))
+        (and (> l start-l) (< l end-l))
+        (and (= l end-l) (<= c end-c)))))
+
+(defun lsp-csharp--code-element-stack-on-l-c (l c elements)
+  "Return omnisharp:CodeElement stack at L (line) and C (column) in ELEMENTS tree."
+  (when-let ((matching-element (seq-find (lambda (el)
+                                           (-let* (((&omnisharp:CodeElement :ranges) el)
+                                                   ((&hash "full" full-range) ranges))
+                                             (lsp-csharp--l-c-within-range l c full-range)))
+                                         elements)))
+    (-let (((&omnisharp:CodeElement :children) matching-element))
+      (cons matching-element (lsp-csharp--code-element-stack-on-l-c l c children)))))
+
+(defun lsp-csharp--code-element-stack-at-point ()
+  "Return omnisharp:CodeElement stack at point as a list."
+  (let ((pos-line (- (line-number-at-pos) 1))
```

Comment:
can you get these from `lsp--cur-position` result? It handles virtual buffers and in the future will handle multi-byte chars in the proper way.